### ISSUE TITLE
Fields are not pickle-able

### ISSUE THREAD

**Issue by ehpor** [ehpor/hcipy#69]
Reported by @chen-pin.

Loading a `Field` object from a fits file is still a `Field`, but doesn't contain the `Field.grid`.

Minimal working example:
```python
from hcipy import *
import numpy as np
import pickle

a = make_pupil_grid(128)
b = Field(np.random.randn(a.size), a)

with open('test_a.pkl', 'wb') as f:
    pickle.dump(a, f)

with open('test_a.pkl', 'rb') as f:
    aa = pickle.load(f)
    print(aa)

with open('test_b.pkl', 'wb') as f:
    pickle.dump(b, f)

with open('test_b.pkl', 'rb') as f:
    bb = pickle.load(f)
    print(bb.grid)
```

Output:
```
CartesianGrid(RegularCoords)
Traceback (most recent call last):
  File ".\hcipy_pickle.py", line 20, in <module>
    print(bb.grid)
AttributeError: 'Field' object has no attribute 'grid'
```

**Comment by ehpor** [comment]
I added a potential fix in the `fix_pickle` branch. The minimal working example above now works as expected. It also passes all existing tests.

@chen-pin can you check if this solves the multiprocessing issues that you encountered?

**Comment by chen-pin** [comment]
Thank you so much!  I will report back later today. 
Sent from Yahoo Mail on Android 
 
  On Sat, Apr 25, 2020 at 7:32 AM, Emiel H. Por<notifications@github.com> wrote:   


I added a potential fix in the fix_pickle branch. The minimal working example above now works as expected. It also passes all existing tests.

@chen-pin can you check if this solves the multiprocessing issues that you encountered?

**Comment by chen-pin** [comment]
I tried the fix, but no luck ...  I first downloaded the fix branch and replaced the <hcipy> directory under <anaconda3> with the downloaded one.  I instantiated a pupil-aperture `Field` and called the `__getstate__` and `__setstate__` methods just to verify the fix is in effect; see below: 
---
>>> x.__getstate__()(1, (1,), dtype('float64'), False, b'\x00\x00\x00\x00\x00\x00\xf0?', <hcipy.field.cartesian_grid.CartesianGrid object at 0x7f14a7775bb0>)
>>> x.__setstate__()                                                         
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __setstate__() missing 1 required positional argument: 'state'---
However, when I run my script, I get the same trace-back:  
---
Traceback (most recent call last):
  File "sleec_run.py", line 148, in <module>
    run_emcee()
  File "sleec_run.py", line 144, in run_emcee
    sampler.run_mcmc(ini_positns, input_pars['nsteps'], progress=True)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/emcee/ensemble.py", line 384, in run_mcmc
    for results in self.sample(initial_state, iterations=nsteps, **kwargs):
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/emcee/ensemble.py", line 285, in sample
    state.log_prob, state.blobs = self.compute_log_prob(state.coords)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/emcee/ensemble.py", line 427, in compute_log_prob
    map_func(self.log_prob_fn, (p[i] for i in range(len(p))))
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/multiprocessing/pool.py", line 768, in get
    raise self._value
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/multiprocessing/pool.py", line 537, in _handle_tasks
    put(task)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/hcipy/optics/optical_element.py", line 713, in attribute
    instance_data = self.get_instance_data(input_grid, output_grid, wavelength)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/hcipy/optics/optical_element.py", line 607, in get_instance_data
    cache_keys = self._get_cache_keys(input_grid, output_grid, wavelength)
  File "/home/pinchen/anaconda3/envs/dev/lib/python3.8/site-packages/hcipy/optics/optical_element.py", line 319, in _get_cache_keys
    raise ValueError('Grid dependent, but no grids are given for lookup.')
ValueError: Grid dependent, but no grids are given for lookup.
---



    On Saturday, April 25, 2020, 7:32:12 AM PDT, Emiel H. Por <notifications@github.com> wrote:  
 
 


I added a potential fix in the fix_pickle branch. The minimal working example above now works as expected. It also passes all existing tests.

@chen-pin can you check if this solves the multiprocessing issues that you encountered?

**Comment by ehpor** [comment]
Ah, the problem was specifically in the agnostic optical elements. The call to get the state of the element was caught by the agnostic operator. I added a fix for this in the same branch. Can you check again?

For documentation purposes, because I'll forget tomorrow... Due to limitations of pickle, lambdas cannot be pickled. Some optical elements (`SurfaceApodizer` for example) require lambdas to operate. `dill` can be used instead of `pickle`, and `multiprocess.Pool` instead of `multiprocessing.Pool`.

My current MWE:
```python
from hcipy import *
import numpy as np
import dill as pickle

pupil_grid = make_pupil_grid(512)
focal_grid = make_focal_grid(4, 16)
elem = FraunhoferPropagator(pupil_grid, focal_grid)

wf = Wavefront(pupil_grid.zeros())
img1 = elem(wf)

with open('test.pkl', 'wb') as f:
    pickle.dump(elem, f)

with open('test.pkl', 'rb') as f:
    elem2 = pickle.load(f)
    img2 = elem2(wf)

    assert np.allclose(img1.electric_field, img2.electric_field)

elem = SurfaceApodizer(pupil_grid.ones(), 1.5)

wf = Wavefront(pupil_grid.zeros())
img1 = elem(wf)

with open('test.pkl', 'wb') as f:
    pickle.dump(elem, f)

with open('test.pkl', 'rb') as f:
    elem2 = pickle.load(f)
    img2 = elem2(wf)

    assert np.allclose(img1.electric_field, img2.electric_field)
```

**Comment by chen-pin** [comment]
It works now!  Thank you so much again!!!  

    On Saturday, April 25, 2020, 6:25:01 PM PDT, Emiel H. Por <notifications@github.com> wrote:  
 
 


Ah, the problem was specifically in the agnostic optical elements. The call to get the state of the element was caught by the agnostic operator. I added a fix for this in the same branch. Can you check again?

For documentation purposes, because I'll forget tomorrow... Due to limitations of pickle, lambdas cannot be pickled. Some optical elements (SurfaceApodizer for example) require lambdas to operate. dill can be used instead of pickle, and multiprocess.Pool instead of multiprocessing.Pool.

My current MWE:
from hcipy import *
import numpy as np
import dill as pickle

pupil_grid = make_pupil_grid(512)
focal_grid = make_focal_grid(4, 16)
elem = FraunhoferPropagator(pupil_grid, focal_grid)

wf = Wavefront(pupil_grid.zeros())
img1 = elem(wf)

with open('test.pkl', 'wb') as f:
    pickle.dump(elem, f)

with open('test.pkl', 'rb') as f:
    elem2 = pickle.load(f)
    img2 = elem2(wf)

    assert np.allclose(img1.electric_field, img2.electric_field)

elem = SurfaceApodizer(pupil_grid.ones(), 1.5)

wf = Wavefront(pupil_grid.zeros())
img1 = elem(wf)

with open('test.pkl', 'wb') as f:
    pickle.dump(elem, f)

with open('test.pkl', 'rb') as f:
    elem2 = pickle.load(f)
    img2 = elem2(wf)

    assert np.allclose(img1.electric_field, img2.electric_field)

**Comment by chen-pin** [comment]
I would like to ask a question about why pull request #70 ("Make OpticalElement and Field pickle-able") does not seem to take effect in the latest master branch.  When I clone the master branch (from https://github.com/ehpor/hcipy ) and install from that, I still get the pickle error.  However, when I use the HCIPy version that was downloaded from #70, no error.  Since #70 has been merged into the master branch, shouldn't the cloned version work without error?Thank you,Pin

    On Sunday, April 26, 2020, 5:26:33 AM PDT, Emiel H. Por <notifications@github.com> wrote:  
 
 


Closed #69 via #70.

**Comment by ehpor** [comment]
That sounds like regression. Apparently the current test code that I wrote doesn't capture your use case. I'll open a new issue to investigate this further.
